openstack_hypervisor/cli/common.py

```
@@ -1,17 +1,154 @@
+SOCKET_FILENAME = "epa.sock"
+SOCKET_PATH = os.path.join(os.environ["SNAP_DATA"], "data", SOCKET_FILENAME)
 click_option_format = click.option(
     "-f",
     "--format",
     default=JSON_FORMAT,
     type=click.Choice([VALUE_FORMAT, JSON_FORMAT, JSON_INDENT_FORMAT]),
     help="Output format",
 )
+
+T = TypeVar("T", bound=BaseModel)
+
+
+class SocketCommunicationError(Exception):
+    """Exception raised when socket communication fails."""
+
+    pass
+
+
+class EPAOrchestratorError(Exception):
+    """Exception raised when the EPA orchestrator returns an error."""
+
+    pass
+
+
+def _communicate_with_socket(
+    request: Union[AllocateCoresRequest, ListAllocationsRequest],
+    response_model: type[T],
+    socket_path: str = SOCKET_PATH,
+) -> T:
+    """Helper function for socket communication with EPA orchestrator.
+
+    Args:
+        request: The request to send to the EPA orchestrator
+        response_model: The expected response model type
+        socket_path: Path to the Unix socket
+
+    Returns:
+        The parsed response from the EPA orchestrator
+
+    Raises:
+        SocketCommunicationError: If socket communication fails (connection,
+                                 data transmission, or validation errors)
+        EPAOrchestratorError: If the EPA orchestrator returns an error response
+    """
+    try:
+        with pysocket.socket(pysocket.AF_UNIX, pysocket.SOCK_STREAM) as s:
+            s.connect(socket_path)
+            s.sendall(request.json().encode())
+            data = s.recv(4096)
+            response_dict = json.loads(data.decode())
+            if "error" in response_dict:
+                raise EPAOrchestratorError(response_dict["error"])
+            response = response_model(**response_dict)
+            return response
+    except (ValidationError, json.JSONDecodeError, pysocket.error, OSError) as e:
+        raise SocketCommunicationError("Socket communication failed: {}".format(e))
+    except EPAOrchestratorError:
```

I haven't seen this pattern, raising an exception, catching and re-raising. Can you take a look and probably easy way is to catch all exceptions. socket only raises 4 exceptions https://docs.python.org/3/library/socket.html#exceptions

---

done